trieste/space.py

```diff
@@ -395,6 +622,146 @@ def sample_sobol(self, num_samples: int, skip: Optional[int] = None) -> TensorTy
             dim=dim, num_results=num_samples, dtype=self._lower.dtype, skip=skip
         ) + self._lower
 
+    def _sample_feasible_loop(
+        self,
+        num_samples: int,
+        sampler: Callable[[], TensorType],
+        max_tries: int = 100,
+    ) -> TensorType:
+        # Rejection sampling using provided callable. Try ``max_tries`` number of times to find
+        # ``num_samples`` feasible points.
+        xs = []
+        count = 0
+        tries = 0
+        while count < num_samples and tries < max_tries:
+            tries += 1
+            xi = sampler()
+            mask = self.is_feasible(xi)
+            xo = tf.boolean_mask(xi, mask)
+            xs.append(xo)
+            count += xo.shape[0]
+
+        if count < num_samples:
+            raise SampleTimeoutError(
+                f"""Failed to sample {num_samples} feasible point(s), even after {tries} attempts.
+                    Sampled only {count} feasible point(s)."""
+            )
+
+        xs = tf.concat(xs, axis=0)[:num_samples]
+        return xs
+
+    def sample_feasible(
+        self, num_samples: int, seed: Optional[int] = None, max_tries: int = 100
+    ) -> TensorType:
+        """
+        Sample feasible points randomly from the space.
+
+        :param num_samples: The number of feasible points to sample from this search space.
+        :param seed: Random seed for reproducibility.
+        :param max_tries: Maximum attempts to sample the requested number of points.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly,
+            from this search space with shape '[num_samples, D]' , where D is the search space
+            dimension.
+        :raise SampleTimeoutError: If ``max_tries`` are exhausted before ``num_samples`` are
+            sampled.
+        """
+        # Without constraints use the normal sample method directly.
```

Comment:
Is there any way to reduce the duplicated code in these methods? Maybe add a `_sample(num_samples, seed, feasible, max_tries)` internal method that gets called from both `sample` and `sample_feasible`?

Reply:
I have had a go in 30a17e7. Please have a look.